### PR TITLE
[SCEV] Teach SCEVExpander to use zext nneg when possible

### PR DESCRIPTION
**Pull request by preames**
zext nneg was recently added to the IR in #67982.  Teaching SCEVExpander to emit nneg when possible is valuable since SCEV may have proved non-trivial facts about loop bounds which would otherwise be lost when materializing the value.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp [hunk 1]
```diff
@@ -1293,7 +1293,10 @@ Value *SCEVExpander::visitTruncateExpr(const SCEVTruncateExpr *S) {
 
 Value *SCEVExpander::visitZeroExtendExpr(const SCEVZeroExtendExpr *S) {
   Value *V = expand(S->getOperand());
-  return Builder.CreateZExt(V, S->getType());
+  auto *Res = Builder.CreateZExt(V, S->getType());
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:1296]: We should add a IsNNeg bool param to IRBuilder::CreateZExt.

Comment (preames) [reply llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:1296]: Agreed.   Once we have a couple examples, a NFC to restructure the API might be called for.

My main point of uncertainty is what we're going to want for the ZExtOrTrunc variants.  Still waiting to see what actual idiomatic usage looks like.

**Comment by nikic** [comment]
There is a test failure in CodeGen/partial_write_in_region_with_loop.ll.

(We should really drop polly from the monorepo...)

**Review by nikic — APPROVED** [review]
LGTM with the test failure fixed.

**Pull request merged**